surplus-deficit/parseData.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * General data munging functionality
+ */
+
+import * as d3 from 'd3';
+import loadData from '@financial-times/load-data';
+
+/**
+ * Parses data file and returns structured data
+ * @param  {String} url Path to CSV/TSV/JSON file
+ * @return {Object}     Object containing series names, value extent and raw data object
+ */
+export function load(url, options) { // eslint-disable-line
+    const { dateFormat } = options;
+    return loadData(url).then((result) => {
+        const data = result.data ? result.data : result;
+        // make sure all the dates in the date column are a date object
+        const parseDate = d3.timeParse(dateFormat);
+        data.forEach((d) => {
+            d.date = parseDate(d.date);
+        });
+        // Automatically calculate the seriesnames excluding the "marker" and "annotate column"
+        const seriesNames = getSeriesNames(data.columns);
+
+        // create stack data object
+        const plotData = d3.stack();
+        plotData.keys(seriesNames);
+
+        plotData.order(d3.stackOrderNone);
+        plotData.offset(d3.stackOffsetNone);
+
+        // Filter data for annotations
+        const annos = data.filter(d => (d.annotate !== '' && d.annotate !== undefined));
+        // Use the seriesNames array to calculate the minimum and max values in the dataset
+        const valueExtent = extentMulti(data, seriesNames);
+        // Format the data that is used to draw highlight tonal bands
+        const boundaries = data.filter(d => (d.highlight === 'begin' || d.highlight === 'end'));
+        const highlights = [];
+
+        boundaries.forEach((d, i) => {
+            if (d.highlight === 'begin') {
+                highlights.push({ begin: d.date, end: boundaries[i + 1].date });
+            }
+        });
+        console.log(highlights);
+
+        return {
+            data,
+            seriesNames,
+            plotData,
+            annos,
+            valueExtent,
+            highlights,
+        };
+    });
+}
+
+
+/**
+ * Returns the columns headers from the top of the dataset, excluding specified
+ * @param  {[type]} columns [description]
+ * @return {[type]}         [description]
+ */
+function getSeriesNames(columns) {
+    const exclude = ['date', 'annotate', 'highlight'];
+    return columns.filter(d => (exclude.indexOf(d) === -1));
+}
+
+/**
+ * Calculates the extent of multiple columns
+ * @param  {[type]} d       [description]
+ * @param  {[type]} columns [description]
+ * @param  {[type]} yMin    [description]
+ * @return {[type]}         [description]
+ */
+// a function that calculates the cumulative ma min values of the dataset
+function getMaxMin(values) {
```

Comment:
This function is neither used nor exported. Please remove if extraneous.

Reply:
Done